### PR TITLE
Add patched package 

### PR DESCRIPTION
**Pull request by sbs2001**
Fixes https://github.com/nexB/vulnerablecode/issues/375

I've done a terrible job at naming things here(see review comments), need help there. 

We would want to move all the intermediate data structures to one place. That will be done in another PR. 

See high level overview at https://github.com/nexB/vulnerablecode/pull/436#issuecomment-823137427

### CODE REVIEW

File: vulnerabilities/tests/test_models.py [hunk 1]
```diff
@@ -66,3 +64,39 @@ def test_vulnerability_save_without_vulnerability_id(self):
                 ).count()
                 == 1
             )
+
+
+# FIXME: The fixture code is duplicated. setUpClass is not working with the pytest mark.
+@pytest.mark.django_db
+class TestPackageRelatedVulnerablity(TestCase):
```
Comment (sbs2001) [inline vulnerabilities/tests/test_models.py:71]: @tdruez  these are the tests

File: vulnerabilities/data_source.py [hunk 2]
```diff
@@ -87,17 +88,16 @@ class Advisory:
 
     summary: str
     vulnerability_id: Optional[str] = None
-    impacted_package_urls: Iterable[PackageURL] = dataclasses.field(default_factory=list)
-    resolved_package_urls: Iterable[PackageURL] = dataclasses.field(default_factory=list)
+    affected_packages_with_patched_package: List[
```
Comment (sbs2001) [inline vulnerabilities/data_source.py:91]: Name problem 1

File: vulnerabilities/helpers.py [hunk 3]
```diff
@@ -20,16 +20,29 @@
...
+from packageurl import PackageURL
+from univers.versions import version_class_by_package_type
 
 # TODO add logging here
 
+is_cve = re.compile(r"CVE-\d{4}-\d{4,7}", re.IGNORECASE).match
+
+
+@dataclasses.dataclass(order=True, frozen=True)
+class AffectedPackageWithPatchedPackage:
```
Comment (sbs2001) [inline vulnerabilities/helpers.py:42]: This will be moved somewhere else in another PR

File: vulnerabilities/helpers.py [hunk 4]
```diff
@@ -100,9 +118,41 @@ def requests_with_5xx_retry(max_retries=5, backoff_factor=0.5):
     return session
 
 
-def contains_alpha(string):
-    """
-    Return True if the input 'string' contains any alphabet
-    """
+def nearest_patched_package(vulnerable_packages, resolved_packages):
```
Comment (sbs2001) [inline vulnerabilities/helpers.py:121]: This is where the magic happens, again it has poor naming. Help.

**Review by pombredanne — CHANGES_REQUESTED** [review]
Some commits message is just "WIP"... may be there is a better message to reword? :angel:

File: vulnerabilities/admin.py [hunk 5]
```diff
@@ -51,7 +51,7 @@ class PackageAdmin(admin.ModelAdmin):
 
 @admin.register(PackageRelatedVulnerability)
 class PackageRelatedVulnerabilityAdmin(admin.ModelAdmin):
-    list_filter = ("is_vulnerable", "package__type", "package__namespace")
+    list_filter = ("package__type", "package__namespace")
```
Comment (pombredanne) [inline vulnerabilities/admin.py:54]: Is there anything that could replace is_vulnerable?

File: vulnerabilities/data_source.py (hunk 2 above)
Comment (pombredanne) [reply vulnerabilities/data_source.py:91]: Could you survey the words use to depict affected/patched in other tools? 
- kaybee/vulas
- dependencycheck and dependency track (and OWASP)
- in general all the data source we use
...
with these listed here we can make a good pick together!

File: vulnerabilities/helpers.py [hunk 6]
```diff
@@ -79,7 +92,12 @@ def create_etag(data_src, url, etag_key):
     return True
 
 
-is_cve = re.compile(r"CVE-\d{4}-\d{4,7}", re.IGNORECASE).match
+def contains_alpha(string):
+    """
+    Return True if the input 'string' contains any alphabet
```
Comment (pombredanne) [inline vulnerabilities/helpers.py:97]:
```suggestion
    Return True if the input 'string' contains any alphanumeric character```

File: vulnerabilities/helpers.py [hunk 7]
```diff
@@ -100,9 +118,41 @@ def requests_with_5xx_retry(max_retries=5, backoff_factor=0.5):
     return session
 
 
-def contains_alpha(string):
-    """
-    Return True if the input 'string' contains any alphabet
-    """
+def nearest_patched_package(vulnerable_packages, resolved_packages):
+    # This class is used to  get around bisect module's lack of supplying custom
```
Comment (pombredanne) [inline vulnerabilities/helpers.py:122]: This should be a docstring

File: vulnerabilities/helpers.py [hunk 8]
```diff
@@ -100,9 +118,41 @@ def requests_with_5xx_retry(max_retries=5, backoff_factor=0.5):
...
+
+    vulnerable_packages = sorted(
+        [PackageURLWithVersionComparator(package) for package in vulnerable_packages]
+    )
+    resolved_packages = sorted(
+        [PackageURLWithVersionComparator(package) for package in resolved_packages]
+    )
+
+    resolved_package_count = len(resolved_packages)
+    affected_package_with_patched_package_objects = []
```
Comment (pombredanne) [inline vulnerabilities/helpers.py:144]: Why the _objects suffix?
```suggestion
    affected_package_with_patched_packages = []
```

File: vulnerabilities/importer_yielder.py [hunk 9]
```diff
@@ -97,13 +97,13 @@
...
         "data_source_cfg": {"repository_url": "https://github.com/RetireNet/Packages.git"},
     },
-    {
-        "name": "suse_backports",
-        "license": "",
-        "last_run": None,
-        "data_source": "SUSEBackportsDataSource",
-        "data_source_cfg": {"url": "http://ftp.suse.com/pub/projects/security/yaml/", "etags": {}},
-    },
+    # {
```
Comment (pombredanne) [inline vulnerabilities/importer_yielder.py:100]: Why is this disabled?

**Review by pombredanne — COMMENTED** [review]
Thanks!
We need to find better names alright and settle on this once and for all IMHO.
vulnerable/affected/resolved/patched are used and I feel we are not consistent.

Also can you explain the high level design here?

**Comment by sbs2001** [comment]
This PR removes the notion of `resolved_package` essentially only storing packages which are vulnerable. It also links these vulnerable packages with nearest(version-wise) package which is safe.

On a lower level, this required :  

1. Removing PackageRelatedVulnerability's `is_vulnerable` flag, since now we only store vulnerable packages primarily.

2. Adding a field called `patched_package` to `PackageRelatedVulnerability` . This is a foreign key to the nearest(wrt `PackageRelatedVulnerability.package` version wise) safe package .

3. Implementing a method called `nearest_patched_package` which accepts 2 lists of vulnerable packages and safe packages (wrt to some same vulnerablity) respectively. This method does the following:

 a. Sort both the lists , ie the packages are now in ascending order wrt their versions. `univers` is used here.
 b. Iterate over each vulnerable package, let's call the package being iterated `pkg`. For each `pkg` we find a package 
               which is just greater(version-wise ) than itself in the safe packages list. See the code [here](https://github.com/sbs2001/vulnerablecode/blob/8e85ff4e75174bbfe0e5169171eafdb81cd2165c/vulnerabilities/helpers.py#L147) 

c. If there exists such a package in (b) we bind this safe package with `pkg` by creating a [AffectedPackageWithPatchedPackage instance](https://github.com/sbs2001/vulnerablecode/blob/8e85ff4e75174bbfe0e5169171eafdb81cd2165c/vulnerabilities/helpers.py#L42)

File: vulnerabilities/importer_yielder.py (hunk 9 above)
Comment (sbs2001) [reply vulnerabilities/importer_yielder.py:100]: We probably want to get rid of that importer, since it only contains `(vulnerability, fixed/backported package)`

File: vulnerabilities/admin.py (hunk 5 above)
Comment (sbs2001) [reply vulnerabilities/admin.py:54]: that should be `patched_package` adding it.

File: vulnerabilities/helpers.py (hunk 8 above)
Comment (sbs2001) [reply vulnerabilities/helpers.py:144]: I'm aware of the principle of not specifying the type. The problem here is this list contains instances of `AffectedPackageWithPatchedPackage` . I don't know how to pluralise that, 

`affected_package_with_patched_packages` -> is it one one affected package with multiple patched packages
`affected_packages_with_patched_package`  -> vice versa 

This is a mess :(

File: vulnerabilities/helpers.py (hunk 3 above)
Comment (sbs2001) [reply vulnerabilities/helpers.py:42]: Use `AffectedPackageWithPatch`

File: vulnerabilities/data_source.py (hunk 2 above)
Comment (pombredanne) [reply vulnerabilities/data_source.py:91]: The general renaming should be tracked in a separate ticket IMHO, not here.

Beyond this, what about `affected_packages` or just `impacted_packages`?

File: vulnerabilities/helpers.py (hunk 3 above)
Comment (pombredanne) [reply vulnerabilities/helpers.py:42]: Or just `AffectedPackage` ?

File: vulnerabilities/helpers.py [hunk 10]
```diff
@@ -20,16 +20,29 @@
...
 
 # TODO add logging here
 
+is_cve = re.compile(r"CVE-\d{4}-\d{4,7}", re.IGNORECASE).match
+
+
+@dataclasses.dataclass(order=True, frozen=True)
+class AffectedPackageWithPatchedPackage:
+    vulnerable_package: PackageURL
+    patched_package: Optional[PackageURL] = None
```
Comment (pombredanne) [inline vulnerabilities/helpers.py:44]: What about: 
```suggestion
    patched_version: Optional[str] = None
```

File: vulnerabilities/helpers.py (hunk 4 above)
Comment (pombredanne) [reply vulnerabilities/helpers.py:121]: we will fix that in a separate ticket.

File: vulnerabilities/helpers.py [hunk 11]
```diff
@@ -100,9 +118,41 @@ def requests_with_5xx_retry(max_retries=5, backoff_factor=0.5):
...
 
-def contains_alpha(string):
-    """
-    Return True if the input 'string' contains any alphabet
-    """
+def nearest_patched_package(vulnerable_packages, resolved_packages):
+    # This class is used to  get around bisect module's lack of supplying custom
+    # compartor. Get rid of this once we use python 3.10 which supports this.
+    # See https://github.com/python/cpython/pull/20556
+    class PackageURLWithVersionComparator:
```
Comment (pombredanne) [inline vulnerabilities/helpers.py:125]: Are you missing a @total_ordering there?

Also you should mention this is only to compare packages of the same type OR you could implement comparison and sorting across types.

Comment (pombredanne) [reply vulnerabilities/helpers.py:125]: You may want to check that the type/ns/name are the same  too

File: vulnerabilities/helpers.py (hunk 8 above)
Comment (pombredanne) [reply vulnerabilities/helpers.py:144]: Based on other review, `affected_package` would be a better name for now.

**Review by pombredanne — APPROVED** [review]
I added  a few more comments for your consideration and please merge.
Let's have a separate ticket for the naming issue.
